OIDC dev services and ui changes (#35324)
quarkusio/quarkus: docs/src/main/asciidoc/security-openid-connect-dev-services.adoc

If you customize `quarkus.http.root-path` or `quarkus.http.non-application-root-path` properties, then replace `q` accordingly.
For more information, see the https://quarkus.io/blog/path-resolution-in-quarkus/[Path resolution in Quarkus] blog post.

== OIDC Lightweight Dev Services

> If you require advanced setup, the Keycloak Dev Services are perfect choice.

Reviewer: It is not really about having an advanced startup, but about Keycloak users preferring to test against the real provider, against real world realms that have to be imported in devmode.

Reviewer 2: I literally copied `advanced setup` it from https://github.com/quarkusio/quarkus/issues/31340 where you wrote `some more advanced realm setup` :-) Nevertheless, I think your suggestion is good an I will fix it. Thanks

Reviewer 2: I re-read https://github.com/quarkusio/quarkus/issues/31340 and I think it tries to suggest that this feature is mainly for non-Keycloak users working in dev mode. Must admit the speed is main advantage in my eyes.

Reviewer 2: Sorry if this sounded too defensive, I just didn't realize difference in our views. I can understand why someone wants to use "a real provider", but as someone who spends most of his live just writing tests, all I want is to have them run quickly, no waiting :-) I can't really leverage starting one container a day and use it all day.

Reviewer: @michalvavrik Np at all, it is just not fair IMHO toward Keycloak users who now depend on the live container starting having to start doing something extra because another group of users are more interested in the start up time savings as opposed to getting anything from the actual live container.

Reviewer 2: I rewrote docs additional entirely, please have a look.

If you customize `quarkus.http.root-path` or `quarkus.http.non-application-root-path` properties, then replace `q` accordingly.
For more information, see the https://quarkus.io/blog/path-resolution-in-quarkus/[Path resolution in Quarkus] blog post.

== Dev Services for OIDC

> The Dev Services for Keycloak discussed so far can be configured to match your Keycloak instance in a production environment.

Reviewer 2: Done.

> For other OpenID Connect providers, it can be more useful to enable the Dev Services for OIDC like in the example below:

Reviewer: ```suggestion
For other OpenID Connect providers, it is recommended to enable the Dev Services for OIDC:
```

Reviewer 2: Done.

[source,properties]
----
quarkus.oidc.devservices.enabled=true
----

Once enabled, Quarkus starts a new server that supports most common OpenID Connect operations.

> Additionally, the Dev Services for OIDC doesn't require container functionality and its startup time outclass any container-based OpenID Connect provider.

Reviewer 2: Done.

If you customize `quarkus.http.root-path` or `quarkus.http.non-application-root-path` properties, then replace `q` accordingly.
For more information, see the https://quarkus.io/blog/path-resolution-in-quarkus/[Path resolution in Quarkus] blog post.

== Dev Services for OIDC

When you work with Keycloak in production, <<dev-services-for-keycloak>> provides the best dev mode experience.
For other OpenID Connect providers, it is recommended to enable the Dev Services for OIDC like in the example below:

[source,properties]
----
quarkus.oidc.devservices.enabled=true
----

> Once enabled, Quarkus starts a new server that supports most common OpenID Connect operations.

Author: s/starts a new server/starts a new OIDC server/

Reviewer 2: Fixed, thanks

Note, the Dev Services for OIDC are enabled by default if Docker and Podman are not available.

> 

Author: Probably we should add 1-2 screenshots and explain that you can log in with alice/alice, bob/bob, or any user name and role list.

For Keycloak we didn't have to do that because it has documentation elsewhere, but for this we should, because nobody knows what it does and how it looks.

Reviewer 2: I have added 2 screenshots and explanation how to log in as alice, bob or any custom user.

Author: This is much better, thanks.

In the future, we should try to find a solution for multi-tenant OIDC too. My most common OIDC demo uses 5+ public OIDC tenants, and I can't use the Dev UI in this case, it's disabled :(

Definitely not in this PR, don't worry :)

If you customize `quarkus.http.root-path` or `quarkus.http.non-application-root-path` properties, then replace `q` accordingly.
For more information, see the https://quarkus.io/blog/path-resolution-in-quarkus/[Path resolution in Quarkus] blog post.

== Dev Services for OIDC

When you work with Keycloak in production, <<dev-services-for-keycloak>> provides the best dev mode experience.
For other OpenID Connect providers, it is recommended to enable the Dev Services for OIDC like in the example below:

[source,properties]
----
quarkus.oidc.devservices.enabled=true
----

NOTE: the Dev Services for OIDC are enabled by default if Docker and Podman are not available.

Once enabled, Quarkus starts a new OIDC server that supports most common OpenID Connect operations.

> You can confirm in the Dev UI console that the OIDC server started, you will see output similar to the following:

Author: I think that's in the regular console too, not just the "Dev UI" console.

Reviewer 2: fixed, thx

[source,shell]
----
2025-01-08 20:50:20,900 INFO  [io.qua.dev.oid.OidcDevServicesProcessor] (build-16) Dev Services for OIDC started on http://localhost:38139
----

> If you navigate to the <<dev-ui-all-oidc-providers>>, you can log into the OIDC server as builtin users `alice` or `bob`:

Author: If you navigate to the DEV UI indeed you can log in, but this is also (in web-app mode) automatic when attempting to access an authenticated REST endpoint.

Reviewer 2: I have added additional information, you might want to re-check. I also deleted couple of lines of KC Dev UI docs, but I am sure @sberyozkin will have plenty to say about it when he is back.

image::dev-ui-oidc-dev-svc-login-page.png[alt=Dev Services for OIDC builtin user login,role="center"]

> As always, default `alice` roles are `admin` and `user`, while default `bob` role is `user`.

Author: ```suggestion
As always, the default roles for `alice`  are `admin` and `user`, while the roles for `bob` are just `user`.
```

Reviewer 2: fixed, thx

Reviewer 2: (note: I can't close comment threads in your PR :-))

> Nevertheless, you can configure roles according to your preference:

Author: ```suggestion
You can configure those built-in roles if required:
```

Reviewer 2: fixed, thx

[source,properties]
----
quarkus.oidc.devservices.roles.alice=root <1>
quarkus.oidc.devservices.roles.bob=guest
----
<1> Assign a `root` role to the user `alice`.

> Another option is log in as a custom user with username and roles of your choice:

Author: ```suggestion
Another option is log in as a custom user with the username and roles of your choice:
```

Reviewer 2: fixed, thx

image::dev-ui-oidc-dev-svc-login-for-custom-users.png[alt=Dev Services for OIDC custom user login,role="center"]

> Whichever user you choose, password is not required.

Author: ```suggestion
Whichever user you choose, no password is required.
```

Reviewer 2: fixed, thx
